### PR TITLE
1. fix research agreement button when you touch it. 2. replacing "civInfo.gameInfo" with "UncivGame.Current.gameInfo" will improve performance.

### PR DESCRIPTION
**Pull request by lishaoxia1985**

### CODE REVIEW

File: core/src/com/unciv/ui/trade/DiplomacyScreen.kt [hunk 1]
```diff
@@ -234,13 +234,13 @@ class DiplomacyScreen(val viewingCiv:CivilizationInfo):CameraStageBaseScreen() {
...
                     val tradeTable = setTrade(otherCiv)
-                    val duration = when(viewingCiv.gameInfo.gameParameters.gameSpeed) {
+                    val duration = when(UncivGame.Current.gameInfo.gameParameters.gameSpeed) {
                         GameSpeed.Quick -> 25
                         GameSpeed.Standard -> 30
                         GameSpeed.Epic -> 45
                         GameSpeed.Marathon -> 90
                     }
-                    val researchAgreement = TradeOffer(Constants.researchAgreement, TradeType.Treaty, duration)
+                    val researchAgreement = TradeOffer(Constants.researchAgreement, TradeType.Treaty, duration, requiredGold)
```
Comment (lishaoxia1985) [inline core/src/com/unciv/ui/trade/DiplomacyScreen.kt]: only this is about fixing Reseach Agreement.

**Comment by r3versi** [comment]
Why it will improve performance? All the game info properties around the code are just references to the same object. I don't see why performance should increase, can you clarify your point?

**Comment by lishaoxia1985** [comment]
i understand no more: 
1. defining a var "gameinfo" in the class civinfo is not necessary.
2. gameinfo should be not changed with civ.
3. if you have a unit, you want use gameinfo in civinfo, you should use unit.civinfo.gameinfo, some example is complicated than this, but using UncivGame.Current.gameInfo is very easy. by the way UncivGame.current is a companion object.
at last, maybe you are right, but UncivGame.Current.gameInfo can be used in some place which we don't know civinfo. we should make this uniform.

File: tests/src/com/unciv/logic/civilization/diplomacy/DiplomacyManagerTests.kt [hunk 2]
```diff
@@ -35,7 +36,7 @@ class DiplomacyManagerTests {
         every { mockGameInfo.getCurrentPlayerCivilization() } returns CivilizationInfo()
 
         for (civ in civilizations.values) {
-            civ.gameInfo = mockGameInfo
+            UncivGame.Current.gameInfo = mockGameInfo
```
Comment (lishaoxia1985) [inline tests/src/com/unciv/logic/civilization/diplomacy/DiplomacyManagerTests.kt]: this edit makes that The Travis CI build failed. I don't know how to solve.😖

**Comment by yairm210** [comment]
Not accepted.
Using companion objects (and any global state) to store state should be avoided whenever possible.
Using the companion objects does not, in fact, improve performance - they are both references, so you're just following the pointers, which is the lowest performance possible
The civinfo should DEFINITELY reference the gameinfo. Civs don't work on a vacuum, they are affected by the game they are in. If anything, accessing a global gameinfo, which may or may not be the same one that the Civ comes from, is a recipe for disaster

**Pull request closed without merging**